source/mainnet/smart-contracts/best-practices/development.rst

```diff
@@ -0,0 +1,203 @@
+Work in iterations, refining the specification when new requirements are discovered or old requirements change.
+
+An informal specification can be refined, turned into a formal one, and used as input to automated testing and verification tools.
+
+The absence of specifications complicates the quality assurance of smart contracts.
+For example, it makes the auditing process less efficient and more expensive, because the specification has to be recovered from the source code before starting the audit process.
+
+
+Concordium Rust Smart Contracts
+===============================
+
+This section provides recommendations for developing smart contracts in Rust.
+
+
+.. _best-practices-code-structure:
+
+Recommended structure
+---------------------
+
+- Use ``cargo init`` with an appropriate template to start a new project.
+- Build contract logic as the state struct implementation.
+
+  .. code-block:: rust
+
+    struct State {
+    ...
+    }
+
+    impl State {
+        fn new() -> Self { ... }
+
+        fn do_something(&mut self, param: MyParameter) { ... }
+
+        ...
+    }
+
+    #[receive(
+        contract = "MyContract",
+        name = "doSomething",
+        parameter = "MyParameter",
+        mutable
+    )]
+    fn contract_do_something<S: HasStateApi>(
+        ctx: &impl HasReceiveContext,
+        host: &mut impl HasHost<State<S>, StateApiType = S>,
+    ) -> ReceiveResult<()> {
+        // Parse parameters
+        let param : MyParameter = ctx.parameter_cursor().get()?;
+        ...
+        // Perform authorization, potentially using using `ctx` info
+        ensure!(sender.matches_account(&owner));
+        ...
+        host.state_mut().do_something(param);
+        ...
+    }
+
+- Fail early: validate input/perform authorization as early as possible in an entrypoint.
+  Returning earlier will save energy and make the call cheaper.
+  Use the ``ensure!()`` macro to validate and return an error.
+
+.. _best-practices-external-calls:
+
+External Calls
+--------------
+
+Every external call should be treated as a potential security risk.
+Calling another contract gives control to potentially malicious code that could make arbitrary calls to any other contract, including your own contract.
+Calls to your contract might change its state through entrypoints that permit updating the state, see :ref:`best-practices-reentrancy`.
+Moreover, you should not make any assumptions about energy consumption, or expect that the execution succeeds.
+You contract should be able to correctly handle situations when the call to an external contract fails.
+
+General recommendations
+^^^^^^^^^^^^^^^^^^^^^^^
+
+- *Avoid complex interactions*. Avoid splitting the on-chain part of your dApp into several smart contracts unless it is strictly necessary.
+  For example, instead of using the *proxy* pattern, use :ref:`natively upgradable contracts <contract-instance-upgradeability>`.
```

Comment:
Please add a link or more description to "the proxy pattern". People might not be aware of this pattern.

Reply:
I'm not sure where to refer. We have an example, but it's in a pull request. Should I reference some Solidity proxy pattern then? Maybe we can just assume if people don't know, then they don't need to know about it anyways. :)

Reply:
I don't like referring to something that users perhaps might not know. We could link to a solidity pattern, mention that it makes everything more complex, and then say that on the superior Concordium blockchain, we can use native upgradability.